File: csharp/DVSClient/DVSClient/address/suggestions/SuggestionsFormatResultFuture.cs
using DVSClient.Exceptions;
using DVSClient.Server.Address.Suggestions;

namespace DVSClient.Address.Suggestions
{
    internal class SuggestionsFormatResultFuture
    {
        private readonly Task<RestApiSuggestionsFormatResponse> _apiFuture;

        public SuggestionsFormatResultFuture(Task<RestApiSuggestionsFormatResponse> apiFuture)
        {
            _apiFuture = apiFuture;
        }

        public async Task<SuggestionsFormatResult> GetAsync()
        {
            try
            {
                var response = await _apiFuture;

                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new SuggestionsFormatResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public async Task<SuggestionsFormatResult> GetAsync(TimeSpan timeout)
        {
            try
            {
                var response = await Task.WhenAny(_apiFuture, Task.Delay(timeout)) == _apiFuture
                    ? await _apiFuture
                    : throw new TimeoutException();

                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new SuggestionsFormatResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException || e is TimeoutException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public bool Cancel(bool mayInterruptIfRunning)
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsCancelled()
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsDone()
        {
            return _apiFuture.IsCompleted;
        }
    }
}

File: java/src/main/java/com/experian/dvs/client/address/suggestions/format/SuggestionsFormatResultFuture.java
package com.experian.dvs.client.address.suggestions.format;

import com.experian.dvs.client.exceptions.EDVSException;
import com.experian.dvs.client.server.address.suggestions.RestApiSuggestionsFormatResponse;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class SuggestionsFormatResultFuture implements Future<SuggestionsFormatResult> {

    private final Future<RestApiSuggestionsFormatResponse> apiFuture;

    public SuggestionsFormatResultFuture(Future<RestApiSuggestionsFormatResponse> apiFuture) {
        this.apiFuture = apiFuture;
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        return this.apiFuture.cancel(mayInterruptIfRunning);
    }

    @Override
    public boolean isCancelled() {
        return this.apiFuture.isCancelled();
    }

    @Override
    public boolean isDone() {
        return this.apiFuture.isDone();
    }

    @Override
    public SuggestionsFormatResult get() throws InterruptedException, ExecutionException {
        final RestApiSuggestionsFormatResponse response = this.apiFuture.get();
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new SuggestionsFormatResult(response);
    }

    @Override
    public SuggestionsFormatResult get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        final RestApiSuggestionsFormatResponse response = this.apiFuture.get(timeout, unit);
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new SuggestionsFormatResult(response);
    }
}
